## Week 8 Lab Instructions:



In [1]:
import pandas
import sklearn
from sklearn.preprocessing import LabelEncoder

from sklearn import preprocessing

from sklearn.ensemble import BaggingClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import train_test_split

# Data location:
# https://archive.ics.uci.edu/ml/machine-learning-databases/adult/

# DF example:


### 1. Return to the US Census Income Dataset from previous weeks
    * We are very familiar with this data at this point.
    * Instantiate the DataFrame in Pandas

In [2]:
df = pandas.read_csv(r'C:\Users\khalil\ML Lab 7\adult.csv')
df.head()

age  workclass  fnlwgt     education  educational-num      marital-status  \
0   25    Private  226802          11th                7       Never-married   
1   38    Private   89814       HS-grad                9  Married-civ-spouse   
2   28  Local-gov  336951    Assoc-acdm               12  Married-civ-spouse   
3   44    Private  160323  Some-college               10  Married-civ-spouse   
4   18          ?  103497  Some-college               10       Never-married   

          occupation relationship   race  gender  capital-gain  capital-loss  \
0  Machine-op-inspct    Own-child  Black    Male             0             0   
1    Farming-fishing      Husband  White    Male             0             0   
2    Protective-serv      Husband  White    Male             0             0   
3  Machine-op-inspct      Husband  Black    Male          7688             0   
4                  ?    Own-child  White  Female             0             0   

   hours-per-week native-country income  
0              40  United-States  <=50K  
1              50  United-States  <=50K  
2              40  United-States   >50K  
3              40  United-States   >50K  
4              30  United-States  <=50K

### 2. Split into training/testing sets

In [12]:
from sklearn.preprocessing import LabelEncoder
df1=df.copy() 
df1=df1.apply(LabelEncoder().fit_transform)
df1.head()

age  workclass  fnlwgt  education  educational-num  marital-status  \
0    8          4   19329          1                6               4   
1   21          4    4212         11                8               2   
2   11          2   25340          7               11               2   
3   27          4   11201         15                9               2   
4    1          0    5411         15                9               4   

   occupation  relationship  race  gender  capital-gain  capital-loss  \
0           7             3     2       1             0             0   
1           5             0     4       1             0             0   
2          11             0     4       1             0             0   
3           7             0     2       1            98             0   
4           0             3     4       0             0             0   

   hours-per-week  native-country  income  
0              39              39       0  
1              49              39       0  
2              39              39       1  
3              39              39       1  
4              29              39       0

In [13]:
x = df1.drop(columns=['age', 'education', 'occupation', 'income'])
y = df1[['income']]


In [14]:
x_train, x_test, y_train, y_test = train_test_split(x, y)

### 3. Build a DecisionTreeClassifier with 5 fold cross validation   
    * Note results 

In [15]:
from sklearn.tree import DecisionTreeClassifier
dt = DecisionTreeClassifier()


In [17]:
results = sklearn.model_selection.cross_val_score(dt, x_train,y_train, cv=5) #5 are number of Folds
results

array([0.80469496, 0.8042588 , 0.8005733 , 0.8039858 , 0.8034398 ])

In [18]:
results.mean(), results.std()

(0.8033905352180225, 0.0014662202047566913)

### 4. Using the bagging technique, build an ensemble with 100 base DecisionTreeClassifier trees
    * Note and compare results 

In [20]:
# Create a bag of estimators of size 100, we use baaging in decision treess
dt_bag = BaggingClassifier(base_estimator=dt, n_estimators=100, n_jobs=4)#4 are number of CORE of machine, 1000 are no of trees

# Fit / Train model
dt_bag.fit(x_train,y_train)

#Results
results = dt_bag.score(x_test, y_test)
results

C:\Users\khalil\anaconda3\lib\site-packages\sklearn\utils\validation.py:63: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(*args, **kwargs)


0.8429285070837769

### 5. Build a RandomForestClassifier   
    * Note and compare results yet again

In [21]:
from sklearn.ensemble import RandomForestClassifier


In [22]:
rf = RandomForestClassifier(n_estimators=100) #100 is no of trees
rf.fit(x_train, y_train)

C:\Users\khalil\AppData\Local\Temp/ipykernel_11952/3369401979.py:2: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  rf.fit(x_train, y_train)


RandomForestClassifier()

In [23]:
rf.score(x_test, y_test)

0.839488985341086

### 6. Lastly, build a different model of your choice.  Compare the results to the best model of the three (rf, bagging, or cv) models built in 1-5.   

In [32]:
#I select KNN Model
clf = KNeighborsClassifier(n_neighbors=5) #5 neighbours

In [33]:
df.dtypes

age                 int64
workclass          object
fnlwgt              int64
education          object
educational-num     int64
marital-status     object
occupation         object
relationship       object
race               object
gender             object
capital-gain        int64
capital-loss        int64
hours-per-week      int64
native-country     object
income             object
dtype: object

In [34]:
clf.fit(x_train, y_train)

C:\Users\khalil\anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:179: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)


KNeighborsClassifier()

In [35]:
clf.predict(x_test)

array([0, 0, 0, ..., 0, 0, 1])

In [36]:
print('Accuracy: {:.2f}'.format(clf.score(x_test, y_test)))

Accuracy: 0.79


As the Accuracy of Bagging-DecisionTreeClassifier(0.84) is Greater than All the other Models Namely: DecisionTreeClassifier-CrossValidation(0.80), RandomForest(0.83), KNN(0.79)